tensorflow_addons/metrics/r_square.py

```diff
@@ -83,6 +96,8 @@ def __init__(
         dtype: AcceptableDTypes = None,
         y_shape: Tuple[int, ...] = (),
         multioutput: str = "uniform_average",
+        penalize: bool = False,
+        num_preds: tf.int32 = 0,
```

Comment:
Why is it set to `0`? By definition, it should be equal to number of independent variables

Reply:
Does num preds arg need to be `tf.int32`?

Reply:
Does num preds arg need to be `tf.int32`?

Reply:
By definition, yes but it would get casted anyways. My point is that this argument represents the number of independent variables used for the model. I don't know why it is set to zero. Check the `p` argument [here](https://en.wikipedia.org/wiki/Coefficient_of_determination#Adjusted_R2)

Reply:
Can you take a look at pytorch lighting https://pytorch-lightning.readthedocs.io/en/latest/_modules/pytorch_lightning/metrics/functional/r2score.html ?

Reply:
I also think that the extra checks on "num_preds" as in pytorch lighting are useful also here.

Reply:
The num_preds should be provided by the user himself,  there's no way of knowing otherwise. I've set it to zero as default because it R2 doesn't require it, only AdjR2. The pytorch_lightning also keeps it to 0 as default.

Reply:
I agree that it should be provided by the user, that's why I am a bit hesitant for a default value for this parameter. Though that is more of a API choice.

Reply:
Now 0 has a specific semantic

Reply:
Worst case scenario: If the user isn't aware of the default argument, then the adjusted r2 score would be much lower, which the end user wouldn't be expecting

Reply:
Setting ```num_preds```  to a value > 0  is the only way to use AdjR2. So, it acts as both a flag and a parameter.

Reply:
I agree. I propose this: If the user didn't provide a value for `num_preds`, then we should raise a warning that the default value of `0` is used. This way the end user would know why the score is much lower than expected. 

This is just my suggestion. Please let me know if you have something better in mind

Reply:
I see your point. The API mainly calculates R2 but allows AdjR2 as an additional(optional) feature. I'm using num_preds as a flag to calculate AdjR2. Incase its default, it falls back to standard R2 :). 

Maybe my naming of ```num_preds``` is a bit off, shall I rename it to ```adjusted``` so that it's more clear that its both a flag and a parameter. Pytorch lightning does the same

Reply:
Another suggestion: Include two parameters, a boolean `adjusted` to indicate whether to calculate adjusted r2 or not. And another `num_features` to indicate the number of independent variables with a default value of `0`. 

```
a warning shouldn't be needed because that's the expected behaviour.
```

IMO, it is needed. Until unless we tell the user that they didn't set the correct value of `num_features`, the calculation will fall back to standard r2 score (the expected behavior)

Reply:
Do we have other use case in TF Addons API where optional parameters are activated only with a dependecy on a another boolean parameter flag?

Reply:
IIRC, `cohen's kappa` has certain bool parameters for specific use cases

Reply:
It is a different from Kappa cause It Is not only a [flag argument case] (https://ardalis.com/are-boolean-flags-on-methods-a-code-smell/) but It will has another argument that depend on the flag one.

Reply:
Okay but is there any problem with the dependency, if at all we include it? I am looking for the `cons` basically

Reply:
It is about a potential dead param that you can change but It is inactivated when the flag is false.

But we could search if this is an anti-pattern or not.

Reply:
E.g. often just the flag parameter is evaluated as code smell

Reply:
I've added warnings for potential incorrect values of num_features, such that it falls back to original R2 score (and there are test to check it). 

Making num_preds both as a flag and param hits two birds with one stone doesn't it? 

Instead of setting "penalize = true" And then check if the user put "num_preds" correctly( num_preds > 0),  we can just have "adjusted" which accounts for both the validity & calculation. As bhack said, a potential dead parameter.

Reply:
Hmm got your point. Thanks for the clarification @bhack and @jonpsy I am okay with it. Just one last thing. Can we rename `num_preds` to `num_features`?

Reply:
Yes it is `number of independent regressors` so probably we could use something better then "num_preds".

Reply:
Sure, I wrote ```num_preds``` because it means "number of predictors". I'm open to a better name tho, anything specific in mind?

Reply:
Sometimes is the number of indipendet regressors/predictor sometimes is "number of explanatory/indipendent variables".
So for me as you like

Reply:
@jonpsy let us know if you have something better in mind. It's the only thing left to do and then we can merge this

Reply:
```num_regressors``` has a nice ring to it :)

Reply:
Let's go with it then